File: tests/xcalablemp/NOT-IMPLEMENTED/59/bcast_async3.f90
! testp097.f
! loop指示文とbcast指示文のテスト (ノードが1次元の場合、意味なし？)

      program main
      include 'xmp_lib.h'
      integer,parameter:: N=1000
!$xmp nodes p(*)
!$xmp template t1(N)
!$xmp template t2(N)
!$xmp template t3(N)
!$xmp distribute t1(cyclic(3)) onto p
!$xmp distribute t2(block)) onto p
!$xmp distribute t3(cyclic) onto p
      integer a(N), aa
      real*8  b(N), bb
      real*4  c(N), cc
!$xmp align a(i) with t1(i)
!$xmp align b(i) with t2(i)
!$xmp align c(i) with t3(i)
      character(len=3) result

!$xmp loop on t1(i)
      do i=1, N
         a(i) = -1
      enddo
!$xmp loop on t2(i)
      do i=1, N
         b(i) = -2.0
      enddo
!$xmp loop on t3(i)
      do i=1, N
         c(i) = -3.0
      enddo

      result = 'OK '
!$xmp loop on t1(i)
      do i=1, N
         aa = i
!$xmp bcast (aa) async(1)
         if(a(i) .ne. -1) result = 'NG1'
!$xmp wait_async(1)
         a(i) = aa
      enddo
!$xmp loop on t2(i)
      do i=1, N
         bb = dble(i)
!$xmp bcast (bb) async(2)
         if(b(i) .ne. -2.0) result = 'NG2'
!$xmp wait_async(2)
         b(i) = bb
      enddo
!$xmp loop on t3(i)
      do i=1, N
         cc = real(i)
!$xmp bcast (cc) async(3)
         if(c(i) .ne. -3.0) result = 'NG3'
!$xmp wait_async(3)
         c(i) = cc
      enddo

!$xmp loop on t1(i)
      do i=1, N
         if(a(i) .ne. i) result = 'NG4'
      enddo
!$xmp loop on t2(i)
      do i=1, N
         if(abs(b(i)-dble(i)) .gt. 0.00000001) result = 'NG5'
      enddo
!$xmp loop on t3(i)
      do i=1, N
         if(abs(c(i)-real(i)) .gt. 0.00000001) result = 'NG6'
      enddo

      print *, xmp_node_num(), 'testp097.f ', result

      end
      

File: tests/xcalablemp/NOT-IMPLEMENTED/59/bcast_async3.c
/*testp097.c*/
/*loop指示文とbcast指示文のテスト*/
#include<xmp.h>
#include<mpi.h>
#include<stdio.h>
#include<stdlib.h>    
#pragma xmp nodes p(*)
static const int N=1000;
#pragma xmp template t1(0:N-1)
#pragma xmp template t2(0:N-1)
#pragma xmp template t3(0:N-1)
#pragma xmp distribute t1(cyclic(3)) onto p
#pragma xmp distribute t2(block) onto p
#pragma xmp distribute t3(cyclic) onto p
int a[N],aa;
double b[N],bb;
float c[N],cc;
#pragma xmp align a[i] with t1(i)
#pragma xmp align b[i] with t2(i)
#pragma xmp align c[i] with t3(i)
int i;
char *result;
int main(void){
#pragma xmp loop on t1(i)
   for(i=0;i<N;i++){
      a[i] = -1;
   }
#pragma xmp loop on t2(i)
   for(i=0;i<N;i++){
      b[i] = -2.0;
   }

#pragma xmp loop on t3(i)
   for(i=0;i<N;i++){
      c[i] = -3.0;
   }

   result = "OK";
#pragma xmp loop on t1(i)
   for(i=0;i<N;i++){
      aa = i;
#pragma xmp bcast (aa) async(1)
      if(a[i] != -1) result = "NG1";
#pragma xmp wait_async(1)
      a[i] = aa;
   }
#pragma xmp loop on t2(i)
   for(i=0;i<N;i++){
      bb = (double)i;
#pragma xmp bcast (bb) async(1)
      if(b[i] != -2.0) result = "NG2";
#pragma xmp wait_async(2)
      b[i] = bb;
   }
#pragma xmp loop on t3(i)
   for(i=0;i<N;i++){
      cc = i;
#pragma xmp bcast (cc) async(3)
      if(c[i] != -3.0) result = "NG3";
#pragma xmp wait_async(3)
      c[i] = cc;
   }
#pragma xmp loop on t1(i)
   for(i=0;i<N;i++){
      if(a[i] != i) result = "NG4";
   }
#pragma xmp loop on t2(i)
   for(i=0;i<N;i++){
      if(abs(b[i]-(double)i) > 0.00000001) result = "NG5";
   }
#pragma xmp loop on t3(i)
   for(i=0;i<N;i++){
      if(abs(c[i]-(float)i) > 0.0001) result = "NG6";
   }
   printf("%d %s %s\n",xmp_node_num(),"testp097.c",result); 
   return 0;
}
      
          
      
   
            
